### PR TITLE
fix(mcu): reduce sample processing delay during active streaming

### PR DESCRIPTION
**Pull request by Jomik**
## Summary

The `AsyncProcessor` batches MCU samples with a 100ms delay before processing on the reactor thread. During active streaming (mesh scanning, probing), this delay risks stale step history when looking up stepper positions. During idle streaming (status, API dump), the delay is fine and reduces reactor wake-ups.

Adds `set_immediate(bool)` to `AsyncProcessor`. MCU `start_streaming`/`stop_streaming` toggle it so position lookups happen as close to sample time as possible during active sessions.

## Decisions & callouts

- The `BATCH_INTERVAL` (0.1s) is unchanged for idle streaming. Only active streaming sessions get immediate dispatch.
- `set_immediate(True)` schedules an immediate callback even if a batched callback is already pending. The duplicate callback is harmless — it finds nothing to process.

### CODE REVIEW

File: src/cartographer/adapters/klipper/mcu/async_processor.py [hunk 1]
```diff
@@ -50,14 +56,43 @@ def __init__(
...
+        without delay. Safe to call multiple times — each call returns
+        a unique token.
+        """
+        with self._lock:
+            self._next_token += 1
+            token = self._next_token
+            self._immediate_tokens.add(token)
+            # Flush pending items on transition to immediate mode
+            if not self._processing_scheduled and self._pending_items:
+                self._processing_scheduled = True
```
[the remark this answers is not in the document]
Comment (Jomik) [reply src/cartographer/adapters/klipper/mcu/async_processor.py]: Fixed — `request_immediate()` now always schedules an immediate callback when pending items exist, regardless of `_processing_scheduled`. The duplicate callback is harmless (finds nothing to process).

File: src/cartographer/adapters/klipper/mcu/mcu.py [hunk 2]
```diff
@@ -223,11 +224,15 @@ def unregister_callback(self, callback: Callable[[Sample], None]) -> None:
...
     def start_streaming(self) -> None:
+        self._immediate_token = self._async_processor.request_immediate()
         self.commands.send_stream_state(enable=True)
 
     @override
     def stop_streaming(self) -> None:
         self.commands.send_stream_state(enable=False)
+        if self._immediate_token is not None:
+            self._async_processor.release_immediate(self._immediate_token)
+            self._immediate_token = None
```
[the remark this answers is not in the document]
Comment (Jomik) [reply src/cartographer/adapters/klipper/mcu/mcu.py]: Fixed — `start_streaming()` now releases the token if `send_stream_state()` throws, and `stop_streaming()` uses `try/finally` to always release.

File: src/cartographer/adapters/klipper/mcu/mcu.py [hunk 3]
```diff
@@ -223,11 +224,15 @@ def unregister_callback(self, callback: Callable[[Sample], None]) -> None:
 
     @override
     def start_streaming(self) -> None:
+        self._immediate_token = self._async_processor.request_immediate()
         self.commands.send_stream_state(enable=True)
 
     @override
     def stop_streaming(self) -> None:
```
[the remark this answers is not in the document]
Comment (Jomik) [reply src/cartographer/adapters/klipper/mcu/mcu.py]: Fixed — `start_streaming()` is now idempotent: returns early if a token is already held.

File: tests/klipper/mcu/test_async_processor.py [hunk 4]
```diff
@@ -353,3 +353,119 @@ def test_no_scheduling_without_items(self) -> None:
...
+
+        # Simulate items queued but callback not yet fired
+        processor.queue_item(3)
+        reactor.callbacks.clear()
+        processor._processing_scheduled = False  # noqa: SLF001  # pyright: ignore[reportPrivateUsage]
+
+        # Request immediate -- should schedule flush for pending item
+        _ = processor.request_immediate()
+        assert len(reactor.callbacks) == 1
+        reactor.run_pending_callbacks()
```
[the remark this answers is not in the document]
Comment (Jomik) [reply tests/klipper/mcu/test_async_processor.py]: Acknowledged — the test is contrived but covers the edge case. Left as-is for now.
